File: src/main/java/fractional_knapsack/fractional_knapsack.java
package fractional_knapsack;

import java.util.Arrays;
import java.util.Comparator;

public class fractional_knapsack {

    public static void main(String[] args) {
        int[] weights = {10, 40, 20, 30};
        int[] values = {60, 40, 100, 120};
        int capacity = 50;

        System.out.println("Maximum value we can obtain = " + getMaxValue(weights, values, capacity));
    }

    private static double getMaxValue(int[] weights, int[] values, int capacity) {
        Item[] iVal = new Item[weights.length];

        for (int i = 0; i < weights.length; i++) {
            iVal[i] = new Item(weights[i], values[i]);
        }

        Arrays.sort(iVal, new Comparator<Item>() {
            @Override
            public int compare(Item o1, Item o2) {
                return o2.cost.compareTo(o1.cost);
            }
        });

        double totalValue = 0.0;
        for (Item i : iVal) {
            int curWt = (int) i.weight;
            int curVal = (int) i.value;

            if (capacity - curWt >= 0) {
                capacity = capacity - curWt;
                totalValue += curVal;
            } else {
                totalValue += (curVal * ((double) capacity / (double) curWt));
                break;
            }
        }

        return totalValue;
    }

    static class Item {
        Double cost;
        double weight, value;

        Item(int weight, int value) {
            this.weight = weight;
            this.value = value;
            cost = (double) value / weight;
        }
    }
}

File: src/main/kotlin/fractional_knapsack.kt

fun main() {
    val weights = listOf(10.0, 40.0, 20.0, 30.0)
    val values = listOf(60.0, 40.0, 100.0, 120.0)
    val capacity = 50.0

    println("Maximum value we can obtain = ${getMaxValue(weights, values, capacity)}")
}

data class KItem(val weight: Double, val value: Double) { val cost = value / weight }

operator fun KItem.plus(other: KItem) = KItem(weight + other.weight, value + other.value)

private fun getMaxValue(weights: List<Double>, values: List<Double>, capacity: Double) =
    values

        .mapIndexed { index, value -> KItem(weights[index], value) }

        .sortedWith(Comparator { c1, c2 -> c2.cost.compareTo(c1.cost) })

        .fold(KItem(0.0, 0.0)) { acc, item ->
            when {
                capacity - acc.weight - item.weight >= 0 ->
                    KItem(item.weight, item.value)
                capacity - acc.weight > 0 ->
                    KItem(capacity, item.value * ((capacity - acc.weight) / item.weight))
                else ->
                    return@fold acc
            } + acc
        }

        .value


inline fun <T, R> Iterable<T>.foldUntil(
    initial: R, operation: (acc: R, T) -> R,
    condition: (acc: R)->Boolean): R {
    var accumulator = initial
    for (element in this) {
        accumulator = operation(accumulator, element)
        if (condition(accumulator)) break
    }
    return accumulator
}











